File: machine_learning_A-Z_template_folder/Part 2 - Regression/Section 5 - Multiple Linear Regression/multiple_linear_regression.R
# Multiple Linear Regression

# Importing the dataset
dataset = read.csv('C:\\Users\\Bill\\Documents\\Training\\Machine Learning A-Z Template Folder\\Part 2 - Regression\\Section 5 - Multiple Linear Regression\\50_Startups.csv')

# Encoding categorical data
dataset$State = factor(dataset$State,
                       levels = c('New York', 'California', 'Florida'),
                       labels = c(1, 2, 3))

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting Multiple Linear Regression to the Training set
#regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
#               data = training_set)

# Fitting Multiple Linear Regression to the Training set -- dot is a shortcut for all independent variables
regressor = lm(formula = Profit ~ .,
               data = training_set)

# Predicting the Test set results
y_pred = predict(regressor, newdata = test_set)
summary(regressor)


#Coefficients:
#  Estimate Std. Error t value Pr(>|t|)    
#(Intercept)      4.965e+04  7.637e+03   6.501 1.94e-07 ***
# R.D.Spend        7.986e-01  5.604e-02  14.251 6.70e-16 ***
# Administration  -2.942e-02  5.828e-02  -0.505    0.617    
# Marketing.Spend  3.268e-02  2.127e-02   1.537    0.134    
# State2           1.213e+02  3.751e+03   0.032    0.974    
# State3           2.376e+02  4.127e+03   0.058    0.954    
---
#  Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

# more stars more significant

# y is the dependent variable -- the rest are independent variables
# P-Value determines the statistical impact of the independent variable on the dependent variable
# ** The LOWER the p-value the more effect the independent variable will have on the dependent variable
  
  
#
# Building the optional model using Backward Elimination
# Remove the statistically insignificant columns < .05
#

SL = 0.05

# Fitting Multiple Linear Regression to the Training set
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
               data = dataset)
summary(regressor)
# Coefficients:
#   Estimate Std. Error t value Pr(>|t|)    
# (Intercept)      5.008e+04  6.953e+03   7.204 5.76e-09 ***
#   R.D.Spend        8.060e-01  4.641e-02  17.369  < 2e-16 ***
#   Administration  -2.700e-02  5.223e-02  -0.517    0.608    
# Marketing.Spend  2.698e-02  1.714e-02   1.574    0.123    
# State2           4.189e+01  3.256e+03   0.013    0.990    
# State3           2.407e+02  3.339e+03   0.072    0.943    
# ---
#   Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

# Remove State

# Fitting Multiple Linear Regression to the Training set
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
               data = dataset)
summary(regressor)

# Coefficients:
#   Estimate Std. Error t value Pr(>|t|)    
# (Intercept)      5.012e+04  6.572e+03   7.626 1.06e-09 ***
#   R.D.Spend        8.057e-01  4.515e-02  17.846  < 2e-16 ***
#   Administration  -2.682e-02  5.103e-02  -0.526    0.602    
# Marketing.Spend  2.723e-02  1.645e-02   1.655    0.105    
# ---
#   Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1


# Remove Administration

# Fitting Multiple Linear Regression to the Training set
regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
               data = dataset)
summary(regressor)

# Coefficients:
#   Estimate Std. Error t value Pr(>|t|)    
# (Intercept)     4.698e+04  2.690e+03  17.464   <2e-16 ***
#   R.D.Spend       7.966e-01  4.135e-02  19.266   <2e-16 ***
#   Marketing.Spend 2.991e-02  1.552e-02   1.927     0.06 .  
# ---
#   Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1


# remove marketing.spend

# Fitting Multiple Linear Regression to the Training set
regressor = lm(formula = Profit ~ R.D.Spend,
               data = dataset)
summary(regressor)

# Coefficients:
#   Estimate Std. Error t value Pr(>|t|)    
# (Intercept) 4.903e+04  2.538e+03   19.32   <2e-16 ***
#   R.D.Spend   8.543e-01  2.931e-02   29.15   <2e-16 ***
#   ---
#   Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1


# Predicting the Test set results
y_pred = predict(regressor, newdata = test_set)

File: machine_learning_A-Z_template_folder/Part 2 - Regression/Section 5 - Multiple Linear Regression/multiple_linear_regression.py
# Multiple Linear Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('C:\\Users\\Bill\\Documents\\Training\\Machine Learning A-Z Template Folder\\Part 2 - Regression\\Section 5 - Multiple Linear Regression\\50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values

# Encoding categorical data
# Encoding the Independent Variable
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:, 3])
onehotencoder = OneHotEncoder(categorical_features = [3])
X = onehotencoder.fit_transform(X).toarray()

# Avoiding the Dummy Variable Trap
X = X[:, 1:]

# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""

# Fitting Multiple Linear Regression to the Training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# Predicting the Test set results
y_pred = regressor.predict(X_test)

#
# Building an optimal model using Backward Elimination
#

# y is the dependent variable -- the rest are independent variables
# P-Value determines the statistical impact of the independent variable on the dependent variable
# ** The LOWER the p-value the more effect the independent variable will have on the dependent variable


# Get the rows and columns
rows, columns = X.shape

# X - intercept 1's
# change b0 to b0X0 ... b1X1
# add add a column of 1's 
# add X matrix to the new matrix of ones -- that way the ones are on column 0
X = np.append(arr = np.ones((rows,1)).astype(int), axis = 1, values = X)

# significance level
SL = 0.05

# All columns
X_opt = X[:,[0,1,2,3,4,5]]

# need stats models library
import statsmodels.formula.api as sm
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

#==============================================================================
                 #coef    std err          t      P>|t|      [95.0% Conf. Int.]
#------------------------------------------------------------------------------
#const       5.013e+04   6884.820      7.281      0.000      3.62e+04   6.4e+04
#x1           198.7888   3371.007      0.059      0.953     -6595.030  6992.607
#x2           -41.8870   3256.039     -0.013      0.990     -6604.003  6520.229
#x3             0.8060      0.046     17.369      0.000         0.712     0.900
#x4            -0.0270      0.052     -0.517      0.608        -0.132     0.078
#x5             0.0270      0.017      1.574      0.123        -0.008     0.062
#==============================================================================

# Remove X2
X_opt = X[:,[0,1,3,4,5]]

# need stats models library
import statsmodels.formula.api as sm
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

#==============================================================================
 #                coef    std err          t      P>|t|      [95.0% Conf. Int.]
#------------------------------------------------------------------------------
#const       5.011e+04   6647.870      7.537      0.000      3.67e+04  6.35e+04
#x1           220.1585   2900.536      0.076      0.940     -5621.821  6062.138
#x2             0.8060      0.046     17.606      0.000         0.714     0.898
#x3            -0.0270      0.052     -0.523      0.604        -0.131     0.077
#x4             0.0270      0.017      1.592      0.118        -0.007     0.061
#==============================================================================
# Remove X1,X2
X_opt = X[:,[0,3,4,5]]

# need stats models library
import statsmodels.formula.api as sm
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

#==============================================================================
                 #coef    std err          t      P>|t|      [95.0% Conf. Int.]
#------------------------------------------------------------------------------
#const       5.012e+04   6572.353      7.626      0.000      3.69e+04  6.34e+04
#x1             0.8057      0.045     17.846      0.000         0.715     0.897
#x2            -0.0268      0.051     -0.526      0.602        -0.130     0.076
#x3             0.0272      0.016      1.655      0.105        -0.006     0.060
#==============================================================================

# Remove X1,X2, X4
X_opt = X[:,[0,3,5]]

# need stats models library
import statsmodels.formula.api as sm
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

#==============================================================================
 #                coef    std err          t      P>|t|      [95.0% Conf. Int.]
#------------------------------------------------------------------------------
#const       4.698e+04   2689.933     17.464      0.000      4.16e+04  5.24e+04
#x1             0.7966      0.041     19.266      0.000         0.713     0.880
#x2             0.0299      0.016      1.927      0.060        -0.001     0.061
#==============================================================================

# Remove X1, X2, X4, X5
X_opt = X[:,[0,3]]

# need stats models library
import statsmodels.formula.api as sm
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
















